### PR TITLE
Guard URI percent decoding bounds

### PR DESCRIPTION
**Pull request by jkwak-work**
## TL;DR

Security vulnerability was reported on the following line:
```
Int charVal = CharUtil::getHexDigitValue(uri[i + 1]) * 16 + CharUtil::getHexDigitValue(uri[i + 2]);
```
This PR adds bounds and validity checks before percent decoding.

## Summary of the problem from the end user perspective

Malformed file URIs from an IDE client could cause the language server to read beyond the URI string while decoding a trailing percent escape, or decode invalid percent escapes into garbage path characters.

## Root cause

`URI::getPath()` decoded every `%` as a three-character escape without first checking that two valid hex digits remained before the path or query boundary.

## Solution in this PR

Only decode a percent escape when both following characters are still within the URI path and are valid hex digits. Too-short or invalid escapes are preserved literally, and local-file URI encoding now emits two-digit hex escapes so encoder/decoder round trips remain consistent.

### Notes to the reviewers; where to focus on

The implementation change is the added guard in `URI::getPath()` plus two-digit padding in `URI::fromLocalFilePath()`. The unit test covers normal `%20` decoding, `%20` at the exact path boundary, trailing `%`, trailing `%2`, invalid high- and low-hex escapes, a `%2` query-boundary case, single-digit-hex local path round trip, and high-bit local path round trip.

### CODE REVIEW

**Comment by jkwak-work** [comment]
@coderabbitai review

**Review by jhelferty-nv — APPROVED** [review]
It would be nice to also have the bot-requested coverage on R172, but this is fine.

File: source/core/slang-io.cpp [hunk 1]
```diff
@@ -1297,7 +1299,7 @@ URI URI::fromLocalFilePath(UnownedStringSlice path)
         else
         {
             char buffer[32];
-            int length = intToAscii(buffer, (int)ch, 16);
+            int length = intToAscii(buffer, (unsigned char)ch, 16, 2);
```
Comment (jhelferty-nv) [inline source/core/slang-io.cpp]: It looks like this is failing in CI because it's choosing an unexpected type for `ch`.

**Review by jhelferty-nv — CHANGES_REQUESTED** [review]

File: source/core/slang-io.cpp (hunk 1 above)
Comment (jkwak-work) [reply source/core/slang-io.cpp]: [Agent] Addressed in 2996c5fee: changed the URI escaping path to pass an int value derived from unsigned char into intToAscii, avoiding intToAscii<unsigned char> and the MSVC warning. Validation: cmake.exe --build --preset debug --target slang-test; build/Debug/bin/slang-test.exe slang-unit-test-tool/uriGetPathPercentDecode.internal.

File: tools/slang-unit-test/unit-test-io.cpp [hunk 2]
```diff
@@ -165,3 +165,13 @@ SLANG_UNIT_TEST(ioLargeFileExists)
     SLANG_IGNORE_TEST
 #endif
 }
+
+SLANG_UNIT_TEST(uriGetPathPercentDecode)
+{
+    SLANG_CHECK(URI::fromString(toSlice("file://path%20name")).getPath() == "path name");
+    SLANG_CHECK(URI::fromString(toSlice("file://path%")).getPath() == "path%");
```
[the remark this answers is not in the document]
Comment (jkwak-work) [reply tools/slang-unit-test/unit-test-io.cpp:174]: [Agent] Addressed in 2996c5fee: added positive boundary cases for a percent escape ending exactly at the path boundary, both at the whole path and before a query string. Validation: cmake.exe --build --preset debug --target slang-test; build/Debug/bin/slang-test.exe slang-unit-test-tool/uriGetPathPercentDecode.internal.

**Comment by jkwak-work** [comment]
@coderabbitai review

File: tools/slang-unit-test/unit-test-io.cpp [hunk 3]
```diff
@@ -165,3 +165,19 @@ SLANG_UNIT_TEST(ioLargeFileExists)
...
 }
+
+SLANG_UNIT_TEST(uriGetPathPercentDecode)
+{
+    SLANG_CHECK(URI::fromString(toSlice("file://path%20name")).getPath() == "path name");
+    SLANG_CHECK(URI::fromString(toSlice("file://%20")).getPath() == " ");
+    SLANG_CHECK(URI::fromString(toSlice("file://a%20?x=1")).getPath() == "a ");
+    SLANG_CHECK(URI::fromString(toSlice("file://path%")).getPath() == "path%");
+    SLANG_CHECK(URI::fromString(toSlice("file://path%2")).getPath() == "path%2");
+    SLANG_CHECK(URI::fromString(toSlice("file://path%2g")).getPath() == "path%2g");
```
[the remark this answers is not in the document]
Comment (jkwak-work) [reply tools/slang-unit-test/unit-test-io.cpp:176]: [Agent] Addressed in d88018b02: added invalid high-hex cases for %g0 and %gg so both halves of the highDigit/lowDigit validity check are covered. Validation: cmake.exe --build --preset debug --target slang-test; build/Debug/bin/slang-test.exe slang-unit-test-tool/uriGetPathPercentDecode.internal.
